### DISCUSSION TITLE
Training and Validation Dice Values are 0

### DISCUSSION THREAD

**Discussion by DexterLuth** [Project-MONAI/MONAILabel#1495] [Q&A]
Hi, 
When training a model all training and validation dice values are zero. configs segmentation module has been tailored to our needs by changing label titles and setting pretrained model loading to false. 
Here is some terminal output: [terminal.txt](https://github.com/Project-MONAI/MONAILabel/files/12099668/terminal.txt)

GPU: NVIDIA GeForce RTX 4090
MONAILabel version: 0.7.0 
Numpy version: 1.24.1 
Pytorch version: 2.1.0.dev20230523+cu117 
MONAILabel rev id: bfb4f8cda6a82d3d6e3d26dbf35980d5ed42f335 

Any help is appreciated!
Thanks!

**Comment by tangy5** [comment]
Yeah, the logs seem all the validation metrics are 0, several items that you can double check:
1. Whether Image-Label pairs are correct, metrics can be 0, if the data is not aligned or values are wrong. Becuase I see the training Dice is also 0, it might be caused by data.
2. I saw the training loss is decreasing in the first 6 epochs, means that the model is learning something. If the loss are correct, you may need to wait a little bit longer to see the metrics, there are digits limit shown in the logs. 
3. If you modified the mtrics, you could double check the metric calculation logics, it could a reason of all 0s.


There could other reasons, hope those helps.

**Reply by DexterLuth** [reply]
Hey tangy5 thanks for the reply!
1. I checked the data and the data is aligned. What do you mean by the fact that the values could be wrong?
2. I have trained the model for longer periods of time and it remains 0, however, I will try again
3. I'm not sure if I modified the metrics, where is the metric calculation logics located?

**Comment by ahmedhosny** [comment]
Hello -  I am hitting the same issue with decreasing loss and zero training and validation dice.
Only changes made to the segmentation model (to train from scratch):

```
 # Labels
 self.labels = {
     "tumor": 1,
 }
 # Download PreTrained Model
 if strtobool(self.conf.get("use_pretrained_model", "false")):
```

**Reply by DexterLuth** [reply]
Hey Ahmed, I think the problem for me was in the data as tangy5 mentioned above, so I would look into that.

**Reply by ahmedhosny** [reply]
Thanks! It happened that I needed 2-3x more samples to get any overlap between predicted and actual segmentations.. and ultimately a non-zero dice.

**Reply by diazandr3s** [reply]
Thanks for the feedback.

By default, MONAI Label uses 20% of the training set for testing. So you'll need at least 4 to 5 annotated samples to see a non-zero Dice score.

If you wish to change this percentage, you could do it in the GUI.

**Comment by Keofire** [comment]
Hello

I have similar problem. The training loss is decreasing, but the dice (and accuracy) remains 0. I did the labels with monailabel-slicer, from zero.
I have ~70 pairs, with consistent labeling.
I ran ~3000 epochs with no success.

I'm lost, don't know what to do.

**Reply by diazandr3s** [reply]
Hi @Keofire,

You may be facing a similar issue as here: https://github.com/Project-MONAI/MONAILabel/discussions/1726

We need to make sure the label indices are consistent throughtout the annotated volumes. Eventhough you created them from scratch, Slicer doesn't consider the label indices (which is key for training the model) - you can usually assign color and name in Slicer, but not indices.

How many regions have you segmented? Is it CT or MR?

Let us know,

CC @lassoan

**Reply by Keofire** [reply]
Thank you for the quick answer!
I used slicer pairing with monailabel, with the "virtual server" running in the background. I thought this ensured the correct labels.

I'm using CT images and trying to do 4 regions.

**Reply by Keofire** [reply]
"Did you resample all the input images to have the same resolution?"

I did nothing like that. Can it be the problem?

**Reply by lassoan** [reply]
You can ensure that segments are consistently mapped to label values by selecting a color table when exporting the segmentation to a labelmap image. You can do it in Slicer, using GUI as described [here](https://slicer.readthedocs.io/en/latest/user_guide/modules/segmentations.html#export-segmentation-to-labelmap-volume-file) or you can use the [`extract_segments` function in slicerio Python package](https://github.com/lassoan/slicerio?tab=readme-ov-file#extract-selected-segments-with-chosen-label-values) in any Python environment.

@diazandr3s it would be great if .seg.nrrd files were not used as is for training but would go through `slicerio.read_segmentation`, `slicerio.extract_segments`, `slicerio.write_segmentation` to ensure that segments are represented by certain label values. Where these could be inserted into the training pipeline?

**Reply by Keofire** [reply]
Now I'm baffled. During labelling, I used the "submit label" button from the monailabel extension in the slicer every time. The "use color table" part is grey, I can't even check it.

My whole process is garbage so far?

**Reply by lassoan** [reply]
Everything is fine. As long as you have set the segment names or terminology consistently for each segment, you can fix up the label values anytime, either manually using GUI or in batch using a few-line Python code snippet.

You can select a color table if you check the "Use color table values" checkbox next to it. Color names in the color table must exactly match the segment names.

**Reply by Keofire** [reply]
Huston we have a problem..


for y in img:
    print(y.shape)
        
(515, 512, 111)
(512, 963, 148)
(512, 563, 66)
.....

**Reply by diazandr3s** [reply]
Hi @Keofire,

Images don't need to be of the same size. However, each pair label/image should be of the same size. Nothing to worry about this if you use Slicer+MONAI Label.

Hope this helps,

**Reply by Keofire** [reply]
Monai label made ".nii.gz" files. 
I'm trying to use slicerio, but I'm already stuck at "slicerio.read_segmentation", it say: "NRRDError: Invalid NRRD magic line. Is this an NRRD file?"

@diazandr3s @lassoan 
Is this maneuver possible with the nifti files?

**Reply by lassoan** [reply]
I've updated slicerio (1.1.0) to be able to load the basic image data from NIFTI files. You then just need to specify the label metadata and write the seg.nrrd file. See example here: https://github.com/lassoan/slicerio?tab=readme-ov-file#create-segmentation-file-from-nifti-labelmap-image-file

**Reply by Keofire** [reply]
@lassoan Thank You very much; I didn't think to be a part of something bigger! I was able to check the data with nibabel and matplotlib but wasn't sure if I could go on with the nifti format. 

Another word on the case: while saving data, I did not tinker with any options. I just used the plain "submit label" function. How is it decided what is the output of the pipeline? I see Monailabel and Slicer as the Swiss army knife of my ages-old plans, there are lots of possibilities to use them!

**Reply by Keofire** [reply]
@lassoan @diazandr3s 

I did as told; I have my segmentations in the final folder, in seg.nrrd format. 

~~I make a server in monailabel (wsl), and start Slicer. I can reach the server, I have my CT scans, but it seems it does not "see" the modified segmentations. In other words: Does Monai label need nifti?~~

~~Update~~
~~Meanwhile, I've been experimenting to see if the problem is on my side. Now I can not even start the monailabel server. Uninstalling-reinstalling even in a fresh Linux environment does not solve my problem. 
ImportError: cannot import name 'AddChanneld' from 'monai.transforms' (/home/keoni/.local/lib/python3.10/site-packages/monai/transforms/__init__.py)~~

**Update2**
Deleting everything (even the radiology app folder) and starting fresh solved the last issue. Now at the startup I can see lines like this:
[WARNING] (monailabel.datastore.local:620) - IGNORE:: No matching image exists for '064147703CT.seg' to add [064147703CT.seg.nrrd]

Does this mean that the chosen format of the segmentation and the images has to match? 
Originally I've started from dicom, converted to nifti. I did this because I've seen the youtube tutorial(s) and that was used there. 



**Update3**
I thought I can win this: I tried to keep the ".nii.gz" and drop the ".seg", but this did not solve the problem.

[2024-08-20 20:14:39,305] [3323] [MainThread] [WARNING] (monailabel.datastore.local:620) - IGNORE:: No matching image exists for '060698904CT.nii.gz' to add [060698904CT.nii.gz.nrrd]

**Reply by Keofire** [reply]
Status report: I tried the single-label route with nifti. It works perfectly! I'm a bit angry of myself not starting this way, I could do all things much-much quicker. 

Do I recommend Monai Label with 3d Slicer to every clinician?  YES
Do I recommend using multi-label? Not yet.

**Reply by diazandr3s** [reply]
Hi @Keofire,

Thanks for your feedback.

Regarding this:

> **Update2**
> Deleting everything (even the radiology app folder) and starting fresh solved the last issue. Now at the startup I can see lines like this:
> [WARNING] (monailabel.datastore.local:620) - IGNORE:: No matching image exists for '064147703CT.seg' to add [064147703CT.seg.nrrd]

> **Update3**
> I thought I can win this: I tried to keep the ".nii.gz" and drop the ".seg", but this did not solve the problem.
> 
> [2024-08-20 20:14:39,305] [3323] [MainThread] [WARNING] (monailabel.datastore.local:620) - IGNORE:: No matching image exists for '060698904CT.nii.gz' to add [060698904CT.nii.gz.nrrd]

Both images and labels should have any of these file formats and not a combination: **_.nii.gz, .nii or .nrrd._**

From what I can see, in Update 2 the issue comes from the fact that you assigned .seg to the image and from Update 3 you assigned .nii.gz.nrrd to the label.

Hope this helps,

**Reply by linfanluntan** [reply]
I got the same zero dice metrics problem in multi-class training, from my tests, I can confirm
1. the training works for image pair with only one single annotated label in data, when multiple annotated labels present in data, training comes with zero dice metrics
2. if load the pre-trained model, training works for image pair with multiple annotated labels present in data.
3. it seems the training stuck at the beginning for image pair with multiple annotated labels present in data if start from scratch.
4. it seems nothing wrong with training data, should be problems of server

Thank you!

**Reply by linfanluntan** [reply]
additionally, I was able to train model from scratch for multi-class segmentation in March of 2024, this month I created a new conda environment and installed monailabel, while the codes are the same as in March, the training is failed on the same training data. The only differences are the new conda environment and new monailabel

**Reply by linfanluntan** [reply]
when installing monailabel it did indicate missing some thing but I can't remember what they are

**Reply by linfanluntan** [reply]
another difference is in pytorch installation, cu113 (on monailabel doc as below) is no more available, and cu118 is used instead
pip install torch torchvision torchaudio --extra-index-url https://download.pytorch.org/whl/cu113

**Reply by lassoan** [reply]
What are the origin, spacing, axis direction of the CT and the segmentation?

CT: in Volumes module, Volume information section
Segmentation: in Data module, MRML node information -> Properties -> Segmentation / Segments / Representations / Binary labelmap / Origin, Spacing, IJKToRASDirections

**Comment by linfanluntan** [comment]


**Reply by linfanluntan** [reply]
![unnamed](https://github.com/user-attachments/assets/64138a8e-f17e-4a71-849b-3099339709d6)

**Reply by linfanluntan** [reply]
![unnamed-1](https://github.com/user-attachments/assets/aaa4b044-b2ae-43ca-939f-71af0430f613)
![unnamed](https://github.com/user-attachments/assets/e3e3023c-1c3f-4216-b39b-46fda3048564)
![unnamed-1](https://github.com/user-attachments/assets/466c365a-eab6-43df-a7f0-09faddc557a8)

**Reply by lassoan** [reply]
You have selected the Scene object in the tree. Please select the segmentation node to get information on the segmentation.

**Reply by lassoan** [reply]
Screenshots are missing. If you have trouble locating the volume information in Volumes and Data module then you can save both the image and the segmentation in NRRD or NHDR format (do not use NIFTI) then you can use any text file viewer to see the image geometry. Please either provide screenshots or copy the NRRD file header here.

**Reply by lassoan** [reply]
OK, I see the segmentation geometry now. Can you copy the CT volume geometry, too?

**Reply by lassoan** [reply]
Well, these don't match.

**CT:**

`space origin: (0,0,0)` and `space directions: (0.91992199420928966,0,0) (0,0.91992199420928966,0) (0,0,5.0000000000000018)` is equivalent to:

Spacing: 0.919922 0.919922 5
IJKToRASDirections:
  -1 0 0
  0 -1 0
  0 0 1
Origin: 0, 0, 0

**Segmentation:**

Spacing: 0.919922 0.919922 5
IJKToRASDirections:
-1 0 0
0 1 0
0 0 1
Origin: 235.04 -235.04 -232.5

Make sure you only use NRRD format and use a MONAI version that includes this fix: https://github.com/Project-MONAI/MONAI/pull/8091

**Reply by lassoan** [reply]
If you cannot figure out from here then I would recommend to create a new issue or discussion where you describe everything from scratch.

**Comment by linfanluntan** [comment]
Ok, I'll check it out.

**Comment by linfanluntan** [comment]
Thanks, I will. BTW, if that means nii.gz won't work?
